### ISSUE TITLE
Does ST_Intersects on albers geometry introduce errors

### ISSUE THREAD

**Issue by benjaminleighton** [CSIRO-enviro-informatics/loci-cache-scripts#23]
Presumably it doesn't because while albers projects doesn't maintain either shape nor length it should equally distort points so overlaps should be preserved

**Comment by dr-shorthair** [comment]
The lines joining adjacent points in a polygonal boundary have different shapes for different projections.  So it is possible that this could introduce overlaps in some cases, especially if the points are widely spaced.

**Comment by benjaminleighton** [comment]
To clarify I mean intersection of two things both projected to albers. I imagine a surface is stretched but never folded so two points that have the relationship prior to albers projection like:
`a --- b` might have a relationship after albers projection like `a -- b` but never like `b -- a`

**Comment by dr-shorthair** [comment]
Yes, I understand. 

If the two things do not share boundary points, then the curves interpolated between them might intersect in one projection, but not in another. I think that is most likely where node-A1 from feature-A falls close to the middle of an interval between two boundary nodes node-B1, node-B2 from the feature-B. The curve joining node-B1 to node-B2 might flip either side of node-A1 depending on the projection.

**Comment by benjaminleighton** [comment]
Looking at the world map example at https://en.wikipedia.org/wiki/Albers_projection I can't see any country boundary that looks like that happens. Shape and length are definitely distorted but all relatively.

**Comment by dr-shorthair** [comment]
It would only likely happen if big enough distances between nodes. But particularly when comparing intersections between a cylindrical projection (e.g. UTM) and a conic projection (e.g. Albers) then there will be cases where this can occur. And I don't think you could call them pathological.

**Comment by benjaminleighton** [comment]
Do you know some lat longs that would illustrate this?

**Comment by dr-shorthair** [comment]
To clarify what I wrote [here](https://github.com/CSIRO-enviro-informatics/loci-cache-scripts/issues/23#issuecomment-538246745): when building a boundary from a sequence of nodes, you join each one to the next using a straight line-segment **in the current projection**. That line-segment will be a curve **in almost any other projection** (particularly if comparing cylindrical vs conic). 

So a point lying on one side of the line in one projection could end up on the other side of a curve in another projection. So for the same set of boundary nodes, the test for point-in-polygon can give a different result when done using different projections.

**Comment by benjaminleighton** [comment]
Yes comparing different projections definitely there is an issue but I'm jut double checking that if we convert two (potentially different) projects to a harmonized representation and then check intersections on that harmonized representation that no errors are introduced. I don't think they are.

**Comment by dr-shorthair** [comment]
It depends on what you mean by 'errors'. There can be different results testing intersection in different projections. Are these errors?

**Comment by benjaminleighton** [comment]
Postgis documents ST_Intersection here http://postgis.net/docs/ST_Intersection.html it works on a range of projections (but probably not across them) it doesn't talk about different kinds of errors.

In particular I mean the case I describe above.

Is there ever a case where if `a--b` in a particular projection then `b--a` will occur when those points are projected to albers. An example lat long should be able to illustrate such a case if it exists

**Comment by dr-shorthair** [comment]
I don't understand `a--b`

**Comment by benjaminleighton** [comment]
Sorry that is my ascii shorthand for two points on earth with a line drawn between them to show they are connected. Perhaps another way of expressing it would be to imagine them as two cities a and b where b lies to the east of a if we projected the location of those cities to albers would there ever be a case where b could now lie to the west of a?

**Comment by dr-shorthair** [comment]
Yes. There will. See argument made in https://github.com/CSIRO-enviro-informatics/loci-cache-scripts/issues/23#issuecomment-538252718 

The question is whether 
```
  a1    b2
     \/
     /\
    b1 \
         \
         a2
```
can change to 
```
 a1   b2
  /  /
 |  /
 |b1  
  \ 
    ---a2
```
and the answer is yes.

**Comment by benjaminleighton** [comment]
I can see how that could happen in the case where in the second image, a1 and a2 are in a different projection to b1 and b2 but I can't see how that could happen if they are all in the same projection. 

Specifically there is a point lets call it c1 in the first projection that lies at the intersection of a1-a2 and b1-b2 and that point must move to a common location in the final diagram.

**Comment by dr-shorthair** [comment]
Yes, both are shown in the same projection now. But the curve joining a1 and a2 in the second case is what a line-segment from a different projection looks like in this projection. So for the same nodes you can get a different result for the intersects() operation depending on the projection used for the geometry computations.

**Comment by benjaminleighton** [comment]
So in the second case lines a and b aren't in the same projection? Presumably though if we project both lines a and lines b we'd end up with something like:

```
 a1   
  /  b2
 |   /
 | /
/  \ 
b1    ---a2
```

and I think I agree that if we did something like ST_Intersects(A(WGS84), B(Albers) we could get a different answer to ST_intersects(A(Albers),B(Albers)).

**Comment by dr-shorthair** [comment]
Node b1 is closer to the line-segment a1-a2 than ascii-art allows. And assume that a1-a2 is a long distance. 

Line-segments in some orientations are more distorted when viewed in other projections than ones in other orientations. 

But what I'm saying is that indeed ST_Intersects(A(WGS84), B(WGS84)) 
could get a different answer to ST_intersects(A(Albers),B(Albers)).

**Comment by dr-shorthair** [comment]
Lets remove b2 and perhaps it will be clearer. 
```
  a1
     \
      \
    b1 \
         \
         a2
```
can change to 
```
 a1  
  / 
 |
 |b1  
  \ 
    ---a2
```

**Comment by dr-shorthair** [comment]
Lets try to clarify: 

1. When you express point positions in a projected system, the coordinates are not lat-lon, they are x-y or easting-northing relative to the origin and axes of the projected coordinate system. They will typically be measured in m or km. So when you draw straight line segments between these points they are straight line segments including all the point on the line between the points _as they would be represented by x-y coordinates in the projected system_. 

2. Are the intersections of polygons defined by the same points on the ground different when computed in different projections? Yes. In the extreme case polygons that clearly intersect in one project do not in another. For example, consider a polygon specified by three points - (Vancouver, UK, Florida) and a second polygon (Greenland). In a polar projection they intersect:
![Polar](https://user-images.githubusercontent.com/608303/66241098-9b8b5100-e741-11e9-8b18-57e0f994631e.png)
Using the Mercator projection they do not:
![Mercator](https://user-images.githubusercontent.com/608303/66241117-a514b900-e741-11e9-897a-219805d25c84.png)
In this case the polygon with lots of vertices (Greenland) is being compared with a polygon of comparable dimensions, but with very few vertices, so it looks like the one moves a lot while the other one doesn't. 

3. But the real issue is what is the surface on which the computations are done. If it is using projected coordinates then in one case the southern tip of greenland is one side of the critical line, and on the other side in another projection. That's if the computations are done using x-y positions in the plane defined by the projection. OTOH if the operation is done on a sphere using lat-lon then you will get a different result. Perhaps you didn't understand that when you are working in a map projection you are no longer using lat-lon, you are using x-y.

**Comment by benjaminleighton** [comment]
I've put together some tests in Postgis to try to illustrate the examples you've given above. 

![image](https://user-images.githubusercontent.com/2150344/66284234-a21cf280-e912-11e9-85f5-5f330cf37582.png)

Is a point in greenland defined in epsg:4326 geodetic datum. 

![image](https://user-images.githubusercontent.com/2150344/66284372-266f7580-e913-11e9-9c0a-10830bb0279d.png)

Is a triangle between Norway Vancouver and Florida defined in epsg:4326. Here it is (likely) being drawn as epsg:3587 (web mercator)

A postgis query like:

```
SELECT ST_Intersects(ST_SetSRID(
ST_Point(-45.648226, 67.673879),4326),
ST_GeomFromText(
'POLYGON((-123.116226 49.246292,12.710824 65.858590,
-81.691069 27.919694, -123.116226 49.246292))'
, 4326));
``` 
returns false 

but a postgis query like:
```
SELECT ST_Intersects(ST_SetSRID(
ST_Point(-45.648226, 67.673879),4326)::geography,
ST_GeomFromText(
'POLYGON((-123.116226 49.246292,12.710824 65.858590,
-81.691069 27.919694, -123.116226 49.246292))'
, 4326)::geography);
``` 
returns true

I think this addresses the concern raised in point 3. 

Postgis currently only provides geography operations for defined geodetic crs like 4326 but it won't work (and perhaps doesn't make sense) with albers map projection. This means that we do have errors introduced because ST_Intersects and ST_Intersection are likely to be working on "straight" line segments defined between polygon points rather. This is unlikely to be an issue at the moment because line segments are normally very short. 

I believe from what you say at the start of point 2 above that we should never see points themselves shift in relative values between projected coordinate reference systems, that is they never change in relative direction to one another. We should never see a relationship between defined points say where one is east of another reverse. This isn't true of line segments which are defined by points they may move depending on how they interpolated.
